### ISSUE TITLE
Copying to a mps tensor produces incorrect results

### ISSUE THREAD

**Issue by Mr4k** [pytorch/pytorch#105277]
### 🐛 Describe the bug

While working on fixing a broken test for `torch.linalg.vander` in #104688, because it should be enabled by cumprod, I noticed that the operator was occasionally inconsistent with the cpu version. I tracked down the inconsistency to a bug in copying to a tensor. After distilling it down here is a small repro:

```
#include <cassert>
#include <torch/torch.h>

torch::Tensor computeResult(torch::DeviceType device) {
    std::cout << "--computing result--\n" << std::endl;
    auto x = torch::arange(2, torch::TensorOptions().device(device));
    std::cout << "x (" << device << ")" << "\n" << x << std::endl;
    auto y = torch::range(0, 3, torch::TensorOptions().device(device)).reshape({2, 2});
    std::cout << "y (" << device << ")" << "\n" << y << std::endl;
    auto x_unsqueeze = x.unsqueeze(1);
    std::cout << "x_unsqueeze (" << device << ")" << "\n" << x_unsqueeze << std::endl;
    y.copy_(x_unsqueeze);
    std::cout << "y after copy" << "\n" << y << std::endl;
    return y;
}

int main() {
    auto cpuResult = computeResult(torch::kCPU);
    auto mpsResult = computeResult(torch::kMPS).to(at::Device(torch::kCPU));
    assert(cpuResult.equal(mpsResult));
    return 0;
}
```

Running the following code I see:
```
--computing result--

x (cpu)
 0
 1
[ CPULongType{2} ]
y (cpu)
 0  1
 2  3
[ CPUFloatType{2,2} ]
x_unsqueeze (cpu)
 0
 1
[ CPULongType{2,1} ]
y after copy
 0  0
 1  1
[ CPUFloatType{2,2} ]
--computing result--

x (mps)
 0
 1
[ MPSLongType{2} ]
y (mps)
 0  1
 2  3
[ MPSFloatType{2,2} ]
x_unsqueeze (mps)
 0
 1
[ MPSLongType{2,1} ]
y after copy
 0  1
 1  3
[ MPSFloatType{2,2} ]
Assertion failed: (cpuResult.equal(mpsResult)), function main, file repro.cpp, line 20.
```

It appears that the copy of the unsqueezed tensor is broken in mps.

### Versions

Collecting environment information...
PyTorch version: 2.1.0.dev20230526
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.3.1 (arm64)
GCC version: Could not collect
Clang version: 14.0.0 (clang-1400.0.29.202)
CMake version: version 3.24.2
Libc version: N/A

Python version: 3.9.12 (main, Apr  5 2022, 01:52:34)  [Clang 12.0.0 ] (64-bit runtime)
Python platform: macOS-13.3.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1 Pro

Versions of relevant libraries:
[pip3] numpy==1.24.3
[pip3] torch==2.1.0.dev20230526
[pip3] torchaudio==2.1.0.dev20230527
[pip3] torchvision==0.16.0.dev20230526
[conda] numpy                     1.24.3                   pypi_0    pypi
[conda] torch                     2.1.0.dev20230526          pypi_0    pypi
[conda] torchaudio                2.1.0.dev20230527          pypi_0    pypi
[conda] torchvision               0.16.0.dev20230526          pypi_0    pypi

cc @kulinseth @albanD @malfet @DenisVieriu97 @razarmehr @abhudev

**Comment by Mr4k** [comment]
Okay I spent a little time today digging further and I think my first example is kind of misleading.

Here's better code:
```
torch::Tensor computeResult(torch::DeviceType device) {
    auto x = torch::arange(2, torch::TensorOptions().device(device));
    auto y = torch::range(5, 10, torch::TensorOptions().device(device)).reshape({2, 3});
    auto x_unsqueeze = x.unsqueeze(1);
    y.copy_(x_unsqueeze);
    return y;
}
```

when run on the cpu this produces:
```
tensor([[0., 0., 0.],
        [1., 1., 1.]])]
```

and when run on mps this produces:
```
tensor([[ 0.,  6.,  7.],
        [ 1.,  9., 10.]], device='mps:0')
```
it looks like the problem is that the tensor does not copy down the entire axis on the mps version. Very important note I didn't notice at first, this occurs when the source tensor is a float and the dst tensor is not. Turns out there is a special path for that in the copy code. I think I've isolated the piece of code where this problem originates based on that path to [here](https://github.com/pytorch/pytorch/blob/73e145532788fd4071a41ea431883bef9633809d/aten/src/ATen/native/mps/operations/Copy.mm#L218)
